File: Graph/dfs_char.cpp
/*
Miêu tả đồ thị:
6 5
A B
A C 
A D
C E
D F
*/
#include <iostream>
#include <unordered_map>
#include <cstring>
#include <vector>
using namespace std;

unordered_map<char, vector<char>> adj;
unordered_map<char, bool> is_visited;

void dfs(char u) {
    cout << u << " ";
    is_visited[u] = true;
    for (char v:adj[u]) {
        if (!is_visited[v]) {
            dfs(v);
        }
    }
}

int main() {
    int n=6, m=5;
    for (char c = 'A'; c < 'A' + n;c++) {
        is_visited[c] = false;
    }
    for (int i=0;i<m;i++) {
        char x, y; cin >> x >> y;
        adj[x].push_back(y);
    }
    dfs('A');
    cout << endl;
    for (const auto& entry:adj) {
        cout << entry.first << " ";
        for (char c : entry.second) {
            cout << c << " ";
        }
        cout << endl;
    }
}

File: Graph/dfs_char.py
## Miêu tả đồ thị
'''
7 13
A B
A C
A F
B C
B D
D A
D C
E C
E G
F A
F C
G D
G E
'''

adj = {}
edges = 13

is_visited = {}

input_data = [
    "A B", "A C", "A F",
    "B C", "B D", "D A",
    "D C", "E C", "E G",
    "F A", "F C", "G D", "G E"
]

for i in range(edges):
    a, b = input_data[i].split()
    is_visited[a] = is_visited[b] = False
    if a not in adj:
        adj[a] = []
    if b not in adj:
        adj[b] = []
    adj[a].append(b)
    # adj[b].append(a)

print(adj)
dfs_list = []
def dfs(adj, u):
    dfs_list.append(u)
    if u not in adj:
        return
    is_visited[u] = True
    for v in adj[u]:
        if not is_visited[v]:
            dfs(adj,v)
dfs(adj,'G')
print(dfs_list)